huntsman/observatory.py

```diff
@@ -512,3 +378,157 @@ def _create_flat_field_observation(self, alt=None, az=None,
+
+            # Stop flats if any time is greater than max
+            exptime_too_long = [t[-1].value >= max_exptime for t in exptimes.values()]
+            if any(exptime_too_long) and not self.past_midnight:
+                self.logger.debug("Exposure times greater than max, stopping flat fields.")
+                break
+
+            # Stop flats if we have taken too many.
+            if any([len(t) >= max_num_exposures for t in exptimes.values()]):
+                self.logger.debug("Too many flats, quitting.")
+                break
+
+            # Handle saturated cameras
+            if is_saturated and exptimes[cam_name][-1].value < min_saturated_exptime:
+                if self.past_midnight:
+                    self.logger.debug('Saturated with short exposure. Stopping flat fields.')
+                    break
+                else:
+                    self.logger.debug('Saturated with short exposure. Waiting 30 seconds'
+                                      'before next exposure.')
+                    max_num_exposures += 1
+                    time.sleep(30)
+
+        # Return the exposure times
+        return exptimes
+
+    def _take_flat_fields(self, exptimes, observation, fits_headers=None, dark=False):
+        """
+        Slew to flat field, take exposures and wait for them to complete.
+        Returns a list of camera events for each camera.
+
+        args:
+            exptimes: dict of camera_name: list of exposure times.
+            observation: Flat field Observation object.
+        """
+        image_dir = self.config['directories']['images']
+        imtype = 'dark' if dark else 'flat'
+        if fits_headers is None:
+            fits_headers = self.get_standard_headers(observation=observation)
+
+        # Slew to field
+        self.logger.debug(f'Slewing to {imtype}-field coords: {observation.field}.')
+        self.mount.set_target_coordinates(observation.field)
+        self.mount.slew_to_target()
+        self.status()  # Seems to help with reading coords
+
+        # Loop over cameras...
+        camera_events = {}
+        for cam_name, exptime in exptimes.items():
+            cam = self.cameras[cam_name]
+
+            # Define filename for flat fields
+            filename = (f'{image_dir}/flats/{cam.uid}/{observation.seq_time}'
+                        f'/{imtype}_{observation.current_exp_num:02d}.{cam.file_extension}')
+
+            # Take exposure and get event
+            exptime = exptime.to(u.second).value
+            camera_event = cam.take_observation(observation, fits_headers,
+                                                filename=filename, exptime=exptime)
+            camera_events[cam_name] = {'event': camera_event, 'filename': filename}
+
+        # Block until done exposing on all cameras
+        while not all([info['event'].is_set() for info in camera_events.values()]):
+            self.logger.debug('Waiting for flat-field images...')
+            time.sleep(1)
```

Comment:
If you're not going to do anything while waiting for these events you can use the `Event.wait()` method and get a timeout mechanism for free:
```suggestion
        for cam_name, info in camera_events.items():
            result = info['event'].wait(timeout=something_sensible)
            if not result:
                raise error.Timeout("Insert message here")
            else:
                self.logger.debug(f"Observation on {cam_name} finished")
```

Reply:
But you probably should also be doing the `safety_func()` check here as this is where the wait is really happening.  However with a sleep of one second that's a little odd.  Maybe need to do something outside the for loop to keep track e.g.:

```py
last_check_time = utils.current_time()
safety_check_interval = 60 # seconds
for cam_name, info in camera_events.items():
    if utils.current_time() - last_check_time > safety_check_interval:
        if not safety_func():
            self.logger.warning('No longer safe)
            break
        last_check_time = utils.current_time()
    
    ...
```

We've always wanted to expand `panoptes.utils.CountdownTimer` to support exactly this kind of countdown/wait with optional callback checks.

Reply:
@wtgee This seems like it could be overkill if we are already checking safety just before we start the exposures. Are we not able to assume that the safety check will be valid for a single exposure?

Reply:
I guess that depends on how long we expect the exposures to be.